File: Algorithms/DynamicProgramming/Examples/GridTraveler/gridTraveler.cpp
#include <iostream>
#include <string>
#include <unordered_map>


std::unordered_map<std::string, long long> memo;

/// @brief Function that calculates the shortest path in a gird traveler
/// @param x size of grid we will be traveling in.
/// @param y size of grid we will be traveling in.
/// @return [int] the amount of ways we can travel grid
auto gridTraveler(int x, int y) -> int 
{
    if (x == 1 && y == 1) { return 1; }
    if (x == 0 || y == 0) { return 0; }

    return gridTraveler(x - 1, y) + gridTraveler(x, y - 1);
}

/// @brief Memoized function that calculates gridTraveler sequence of a given number
/// @param x size of grid we will be traveling in.
/// @param y size of grid we will be traveling in.
/// @return [Long] the resulting gridTraveler number
auto gridTravelerMemoized (int x, int y) -> long long
{
    std::string key = std::to_string(x) + "," + std::to_string(y);

    if (memo.count(key)) { return memo[key]; }
    if (x == 1 && y == 1) { return 1; }
    if (x == 0 || y == 0) { return 0; }

    long long result = gridTravelerMemoized(x - 1, y) + gridTravelerMemoized(x, y - 1);
    memo[key] = result;
    return result;
}

/// @brief Main Code
auto main() -> int
{
    std::cout << "------------------ NON MEMOIZED CODE ------------------ " << std::endl;
    std::cout << "Amount of paths for {1, 1} is: [" << gridTraveler(1, 1) << "]" << std::endl;
    std::cout << "Amount of paths for {2, 3} is: [" << gridTraveler(2, 3) << "]" << std::endl;
    std::cout << "Amount of paths for {3, 2} is: [" << gridTraveler(3, 2) << "]" << std::endl;
    std::cout << "Amount of paths for {3, 3} is: [" << gridTraveler(3, 3) << "]" << std::endl;

    std::cout << "-------------------- MEMOIZED CODE -------------------- " << std::endl;
    std::cout << "MEMOIZED Amount of paths for {18, 18} is: [" << gridTravelerMemoized(18, 18) << "]" << std::endl;
    memo.clear();

    return 1;
}

File: Algorithms/DynamicProgramming/Examples/GridTraveler/gridTraveler.js
/** ---------------------------------------------
 * 
 * Memoize a Grid Traveler function
 --------------------------------------------- */

/**
 * Function that calculates the shortest path in a gird traveler
 * 
 * @param {Number} x size of grid we will be traveling in.
 * @param {Number} y size of grid we will be traveling in.
 * @returns [Number] the amount of ways we can travel grid
 */
const gridTraveler = (x, y) => {
    if (x === 1 && y === 1) { return 1; }
    if (x === 0 || y === 0) { return 0; }

    return gridTraveler(x - 1, y) + gridTraveler(x, y - 1);
}

/**
 * Memoized function that calculates gridTraveler sequence of a given number
 * 
 * @param {Number} x size of grid we will be traveling in.
 * @param {Number} y size of grid we will be traveling in.
 * @param {Object} memo used as a cache to store previously computed values.
 * @returns [Number] the resulting gridTraveler number
 */
const gridTravelerMemoized = (x, y, memo = {}) => {
    const key = x + ',' + y;

    if (key in memo) { return memo[key]; }
    if (x === 1 && y === 1) { return 1; }
    if (x === 0 || y === 0) { return 0; }

    memo[key] = gridTravelerMemoized(x - 1, y, memo) + gridTravelerMemoized(x, y - 1, memo);
    return memo[key];
}

/**
 * Tabulated Function that calculates the shortest path in a gird traveler
 * O(n) time complexity           O(n) Space complexity
 * @param {Number} x size of grid we will be traveling in.
 * @param {Number} y size of grid we will be traveling in.
 * @returns [Number] the amount of ways we can travel grid
 */
const gridTravelerTabulated = (x, y) => {
    const table = Array(x + 1)
        .fill()
        .map(() => Array(y + 1).fill(0));
    
    table[1][1] = 1;

    for (let i = 0; i <= x; i++) {
        for (let j = 0; j <= y; j++) {
            const current = table[i][j];
            if (j + 1 <= x) { table[i][j + 1] += current; }
            if (i + 1 <= y) { table[i + 1][j] += current; }
        }
    }

    return table[x][y]
}

// Main Code

console.log("------------------ NON MEMOIZED CODE ------------------ ");
console.log(`Amount of paths for {1, 1} is: [${gridTraveler(1, 1)}]`);
console.log(`Amount of paths for {2, 3} is: [${gridTraveler(2, 3)}]`);
console.log(`Amount of paths for {3, 2} is: [${gridTraveler(3, 2)}]`);
console.log(`Amount of paths for {3, 3} is: [${gridTraveler(3, 3)}]`);

console.log("-------------------- MEMOIZED CODE -------------------- ");
console.log(`MEMOIZED Amount of paths for {18, 18} is: [${gridTravelerMemoized(18, 18)}]`);

console.log("------------------- TABULATED CODE --------------------");
console.log(`TABULATED Amount of paths for {18, 18} is: [${gridTravelerTabulated(18, 18)}]`);